src/xrpld/app/misc/CredentialsHelper.cpp

```diff
@@ -0,0 +1,204 @@
+
+#include <xrpld/app/misc/CredentialsHelper.h>
+#include <xrpld/ledger/View.h>
+
+namespace ripple {
+namespace credentials {
+
+bool
+checkExpired(
+    std::shared_ptr<SLE const> const& sleCredential,
+    NetClock::time_point const& closed)
+{
+    std::uint32_t const exp = (*sleCredential)[~sfExpiration].value_or(
+        std::numeric_limits<std::uint32_t>::max());
+    std::uint32_t const now = closed.time_since_epoch().count();
+    return now > exp;
+}
+
+bool
+removeExpired(ApplyView& view, STTx const& tx, beast::Journal const j)
+{
+    auto const closeTime = view.info().parentCloseTime;
+    bool foundExpired = false;
+
+    STVector256 const& arr(tx.getFieldV256(sfCredentialIDs));
+    for (auto const& h : arr)
+    {
+        // Credentials already checked in preclaim. Look only for expired here.
+        auto const k = keylet::credential(h);
+        auto const sleCred = view.peek(k);
+
+        if (checkExpired(sleCred, closeTime))
+        {
+            JLOG(j.trace())
+                << "Credentials are expired. Cred: " << sleCred->getText();
+            // delete expired credentials even if the transaction failed
+            deleteSLE(view, sleCred, j);
+            foundExpired = true;
+        }
+    }
+
+    return foundExpired;
+}
+
+TER
+deleteSLE(
+    ApplyView& view,
+    std::shared_ptr<SLE> const& sleCredential,
+    beast::Journal j)
+{
+    if (!sleCredential)
+        return tecNO_ENTRY;
+
+    auto delSLE =
+        [&view, &sleCredential, j](
+            AccountID const& account, SField const& node, bool isOwner) -> TER {
+        auto const sleAccount = view.peek(keylet::account(account));
+        if (!sleAccount)
+        {
+            JLOG(j.fatal()) << "Internal error: can't retrieve Owner account.";
+            return tecINTERNAL;
+        }
+
+        // Remove object from owner directory
+        std::uint64_t const page = sleCredential->getFieldU64(node);
+        if (!view.dirRemove(
+                keylet::ownerDir(account), page, sleCredential->key(), false))
+        {
+            JLOG(j.fatal()) << "Unable to delete Credential from owner.";
+            return tefBAD_LEDGER;
+        }
+
+        if (isOwner)
+            adjustOwnerCount(view, sleAccount, -1, j);
+
+        return tesSUCCESS;
+    };
+
+    auto const issuer = sleCredential->getAccountID(sfIssuer);
+    auto const subject = sleCredential->getAccountID(sfSubject);
+    bool const accepted = sleCredential->getFlags() & lsfAccepted;
+
+    auto err = delSLE(issuer, sfIssuerNode, !accepted || (subject == issuer));
+    if (!isTesSuccess(err))
+        return err;
+
+    if (subject != issuer)
+    {
+        err = delSLE(subject, sfSubjectNode, accepted);
+        if (!isTesSuccess(err))
+            return err;
+    }
+
+    // Remove object from ledger
+    view.erase(sleCredential);
+
+    return tesSUCCESS;
+}
+
+NotTEC
+check(PreflightContext const& ctx)
```

Comment:
This method should be called `preflightCheck` or something - `check` is a bit too vague.

Reply:
`check` and `valid` are kind of confusing since their names make them seem interchangeable.

Reply:
These names will be used as `credentials::check() / credentials::valid` and agreed with Greg. Because the names of these functions have changed too many times feel free to start slack discussion to agreed on final variant.

Reply:
After discussion, can this be renamed to `checkFields`?

Reply:
renamed